### ISSUE TITLE
Finer-grained cost accounting

### ISSUE THREAD

**Issue by shlevy** [nixbuild/feedback#12]
It would be great if cost accounting could be broken down by some sort of hierarchical metadata associated with a given session (perhaps tags, or perhaps some new mechanism just for this). E.g. I want to be able to see "how many build hours did my account use", "how many build hours did the `foo` team on my account user", "how many build hours did the `foo` team on my account use for the `bar` project".

One possible design would be that builds could be annotated with a cost group, and cost groups can be nested (so `[ foo ]` would be the cost group of the `foo` team in my example, while `[ foo, bar ]` would be that used by `foo` on the `bar` project). Then any request for the totals for a cost group would include any sub-groups.

Possible further functionality:

- Ability to require some key to have its builds associated with some cost group or its descendents
- Ability to set quotas by cost group (perhaps per-key)
- Some kind of split accounting: A build could be annotated with multiple groups, and/or a group could have multiple parents, and there's some policy describing how the time is divided up between the groups (could be "each group gets the whole number" or "this group gets half, that one gets a quarter, that one gets a quarter").

**Comment by rickynils** [comment]
You can use tags combined with the [/builds/summary](https://docs.nixbuild.net/api/#get-buildssummary) API endpoint to _track_ build time usage today. However, there is nothing that forces a user to set specific tags on their builds (this could be solved by adding build tags to the Biscuit context), and there is nothing in place for enforcing build time limits/quotas as described.

**Comment by shlevy** [comment]
If the total time usage by tag were available as a fact at verification then we could put the policy in place in the token as well.

**Comment by rickynils** [comment]
Yeah, the only difficulty would be to know which tag (or combination of tags) that should be used, since we can't add all possible tag-times as facts.

**Comment by shlevy** [comment]
Oh, there's no way to dynamically populate facts based on what's actually requested in the policy? I wonder how hard that would be to add to the validation lib... :sob:

**Comment by rickynils** [comment]
Facts added by attenuators can't be used by the authorizer. So you would have to add such facts during the initial token creation. I do think there is some work in Biscuit to change this, though, but I'm not sure where I saw that.

An alternative would to populate the facts based on the request. So you could have some meta-tag on your builds for telling us which tags you want to track.

**Comment by rickynils** [comment]
@shlevy My last suggestion, to have some way to set the "accounting group" per-build (through the SSH environment), would it be enough for your needs, or too simplistic? If we had that, you could define policies that required a specific group to be used, and also define limits (since we could add something like "the last 30 days spent CPU time  for the group" as a Biscuit fact).

**Comment by shlevy** [comment]
Yes, that should work!
